maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java

```diff
@@ -238,9 +268,82 @@ else if ( request.isUpdateSnapshots() )
         mavenRepositorySystem.injectProxy( session, request.getPluginArtifactRepositories() );
         mavenRepositorySystem.injectAuthentication( session, request.getPluginArtifactRepositories() );
 
+        if ( Boolean.getBoolean( "maven.experimental.buildconsumer" ) )
+        {
+            session.setFileTransformerManager( newFileTransformerManager() );
+        }
         return session;
     }
 
+    private FileTransformerManager newFileTransformerManager()
+    {
+        return new FileTransformerManager()
+        {
+            @Override
+            public Collection<FileTransformer> getTransformersForArtifact( final Artifact artifact )
+            {
+                Collection<FileTransformer> transformers = new ArrayList<>();
+                if ( "pom".equals( artifact.getExtension() ) )
+                {
+                    final TransformerFactory transformerFactory = Factories.newTransformerFactory();
+
+                    transformers.add( new FileTransformer()
+                    {
+                        @Override
+                        public InputStream transformData( File file )
+                            throws IOException, TransformException
+                        {
+                            final PipedOutputStream pipedOutputStream  = new PipedOutputStream();
+                            final PipedInputStream pipedInputStream  = new PipedInputStream( pipedOutputStream );
+                            
+                            final SAXSource transformSource;
+                            try
+                            {
+                                transformSource =
+                                    new SAXSource( consumerPomXMLFilterFactory.get().get( file.toPath() ),
+                                                   new InputSource( new FileReader( file ) ) );
+                            }
+                            catch ( SAXException | ParserConfigurationException | TransformerConfigurationException e )
+                            {   
+                                throw new TransformException( "Failed to create a consumerPomXMLFilter", e );
+                            }
+                            
+                            final StreamResult result = new StreamResult( pipedOutputStream );
+                            
+                            final Callable<Void> callable = () ->
+                            {
+                                try ( PipedOutputStream out = pipedOutputStream )
+                                {
+                                    transformerFactory.newTransformer().transform( transformSource, result );
+                                }
+                                return null;
+                            };
+
+                            ExecutorService executorService = Executors.newSingleThreadExecutor();
```

Comment:
@rmannibucau it seems like this implementation can cause deadlocks (as mentioned in the PipedInputStream/PipedOutputStream Javadocs). I see this happening when trying to active this feature by default, some unittests start to hang. Anyone knows how to solve this?

Reply:
Do you need a thread at all? Seems you sublit the task to wait for it, why not just chaining transformers streams? Swizzle does that with success.

Reply:
Yes, please read https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/io/PipedInputStream.html.
I don't want to buffer the streams, that can consume too much memory. And it is quite easy to get the deadlock: fork, switch to this branch, change default value in Features to `true`  and run `mvn test`.